jipt/src/ui/pin.ts

```diff
@@ -0,0 +1,82 @@
+import LiveNode from '../mutation/live-node';
+import State from '../state';
+import styles from './styles';
+import UI from './ui';
+
+interface Props {
+  ui: UI;
+  liveNode: LiveNode;
+  root: Element;
+  state: State;
+}
+
+/*
+  The Pin component serves as the entrypoint for the user. The element it creates
+  is responsible to sending messages to the Accent UI.
+*/
+export default class Pin {
+  private readonly element: Element;
+  private readonly liveNode: LiveNode;
+  private readonly state: State;
+  private readonly ui: UI;
+
+  constructor(props: Props) {
+    this.state = props.state;
+    this.liveNode = props.liveNode;
+    this.ui = props.ui;
+
+    const pin = document.createElement('div');
+    styles.hide(pin);
+
+    props.root.append(pin);
+
+    this.element = pin;
+  }
+
+  bindEvents() {
+    this.element.addEventListener('click', event => {
+      const target = event.target as HTMLElement;
+
+      if (target.dataset.id) {
+        this.ui.selectTranslation(target.dataset.id);
+        styles.hide(this.element);
+      }
+      if (target.dataset.ids) {
+        this.ui.selectTranslations(target.dataset.ids);
+        styles.hide(this.element);
+      }
+    });
+
+    document.addEventListener('mouseover', event => {
+      const node = event.target as HTMLElement;
+
+      if (this.liveNode.isLive(node)) this.showFor(node);
+    });
+  }
+
+  private showFor(target: HTMLElement) {
+    const {x, y, height} = target.getBoundingClientRect() as DOMRect;
```

Comment:
I wonder why you need to cast it here 🤔

Reply:
Oh ok, thats weird, it returns a union type of `ClientRect | DOMRect`.

```ts
interface ClientRect {
    bottom: number;
    readonly height: number;
    left: number;
    right: number;
    top: number;
    readonly width: number;
}

interface DOMRect extends DOMRectReadOnly {
    height: number;
    width: number;
    x: number;
    y: number;
}
```

The reason seem to be [a compatibility issue](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect#Notes):

>The returned DOMRect object can be modified in modern browsers. This was not true with older versions which effectively returned DOMRectReadOnly. With IE and Edge, not being able to add missing properties to their returned ClientRect, object prevents backfilling x and y.
>
>Due to compatibility problems (see below), it is safest to rely on only properties left, top, right, and bottom.

I think it would be best to rely on top/left instead of x/y.